semgrep-core/toy-matcher/Matcher.ml

```diff
@@ -0,0 +1,562 @@
+(*
+   A matcher operating on sequences of symbols,
+   using the same basic matching algorithm as semgrep-core.
+
+   The goals are:
+
+   - Clarify semgrep-core's matching algorithm by not having to deal with any
+     of the real-world complexity.
+   - Keep the semgrep-core implementation separate and easy to follow,
+     without functors.
+*)
+
+open Printf
+
+type pattern_atom =
+  | Any_symbol         (* match any single symbol *)
+  | Symbol of char     (* match a specific symbol *)
+  | Ellipsis           (* match any sequence of symbols *)
+  | Backref of string  (* match the same thing that was previous matched
+                          by the pattern atom of that name. *)
+
+(*
+   A pattern is a list of pattern atoms with an optional name.
+   The presence of a name indicates that the symbol or sequence it matches
+   must be captured, i.e. stored in the environment.
+*)
+type pattern = (pattern_atom * string option) list
+
+type input = char list
+
+(* Captured subsequences. *)
+module Env = Map.Make (String)
+type env = char list Env.t
+
+(*
+   Accumulator use to store the symbols matched by the current ellipsis.
+
+     None = we're not in an ellipsis
+     Some None = we're in an ellipsis but it has no name and it's not capturing
+
+     Some (Some ("foo", acc)) = acc is the stack of symbols, i.e. the list
+     of symbols captured so far by the current ellipsis in reverse order
+*)
+type ellipsis = (string * char list) option option
```

Comment:
maybe better to define a new type for those 3 variants instead of using option option.
type capture_ellipsis = NotInEllipsis | EllipsisButNoCapture | EllipsisCapture of string * char list

Reply:
I did that.